### ISSUE TITLE
x/tools, x/build: create a relui workflow for gopls release automation

### ISSUE THREAD

**Issue by dle8** [golang/go#57643]
This issue tracks the relui workflow to automate gopls release namely:

- [ ] Create an issue to track the release
- [ ] Creating a new release branch
- [ ] CL to update the release branch from master
- [ ] CL to update the debug version
- [ ] CL to update gopls/go.mod to require the correct x/tools
- [ ] Tag the prerelease
- [ ] Generate the release note
- [ ] Create a new release in github
- [ ] Tag a final release
- [ ] Update dependencies for x/tools and x/tools/gopls

CC @findleyr

**Comment by h9jiang** [comment]
In the next change, we will add a step to create a branch for a minor version (X.Y.0) release. 

In my local run, the release flow failed because of permission denied issue when trying to create the next minor release branch `gopls-release-branch.0.17`. (WAI) The permission issue should not happen in prod.

![image](https://github.com/user-attachments/assets/82c72761-1c22-4d07-831c-e066a330b4ec)

**Comment by h9jiang** [comment]
In the next change, we will add a step to change the `codereview.cfg` under the root dir of `x/tools` repo, and create a CL based on the change. The flow will take `reviewers` as optional parameters. 

In my local run, I short circuit the previous steps (i.e. version validation, branch creation). Provide input parameter `version = v0.1.0`. The `codereview.cfg` in release branch `gopls-release-branch.0.1` is not in desired state yet. 

So the flow will create a CL to update `codereview.cfg` accordingly. ( Reconcile it to desired state. :D )

![image](https://github.com/user-attachments/assets/8e265403-9d2b-4082-97de-80a712a75d48)

![image](https://github.com/user-attachments/assets/10ef5de7-ec26-4268-be38-c144403ff5d2)

**Comment by h9jiang** [comment]
In the next change, we will add a step to update the `x/tools` dependency in the `gopls` sub module, and create a CL based on the change.

The way of updating the dependency: 
* Drop the replace directive of `x/tools` so the head of release branch will not point to its parent. 
* Pin the `x/tools` to the head of the current release branch.

In my local run, I short circuit the previous steps (i.e. branch creation). Provide input parameter version = `v0.16.2`. The `go.mod & go.sum` in release branch `gopls-release-branch.0.16` will be updated to the head of the release branch.

The CL created update the `go.mod`

```
- golang.org/x/tools v0.22.1-0.20240628205440-9c895dd76b34
+ golang.org/x/tools v0.22.1-0.20240715174654-66fa8e7351ea
```

And the `git log -1`

```
commit 66fa8e7351eaa1909b6afcaa95e7adb4e08b1714 (HEAD -> gopls-release-branch.0.16, origin/gopls-release-branch.0.16)
Author: Rob Findley <rfindley@google.com>
Date:   Fri Jul 12 21:56:01 2024 +0000

    [gopls-release-branch.0.16] gopls: update x/telemetry dependency
    
    Update x/telemetry to pick up recent bug fixes (notably
    golang/go#68311).
    
    Change-Id: Ic95d54fc43872ab5c3548cf47134fbaea27e1487
    Reviewed-on: https://go-review.googlesource.com/c/tools/+/598115
    Reviewed-by: Dmitri Shuralyov <dmitshur@google.com>
    Reviewed-by: Dmitri Shuralyov <dmitshur@golang.org>
    LUCI-TryBot-Result: Go LUCI <golang-scoped@luci-project-accounts.iam.gserviceaccount.com>
    (cherry picked from commit 6f4e2a811e8b599d25464bc036a8abda0088d4ed)
    Reviewed-on: https://go-review.googlesource.com/c/tools/+/598056
    Auto-Submit: Robert Findley <rfindley@google.com>
```

The CL change the version of `x/tools` in `gopls` submodule to the head of this release branch. 

![image](https://github.com/user-attachments/assets/043283d8-69bf-492b-9f7b-4b7579fcbb1a)
![image](https://github.com/user-attachments/assets/97cd6db6-982a-4bd7-9a1c-123e9ced2c69)

**Comment by h9jiang** [comment]
In the next change, we will add a step to verify we can successfully install gopls from the head of the branch using the `go install`golang.org/x/tools/gopls@{BRANCH_NAME}`.

![image](https://github.com/user-attachments/assets/8f8822fe-ba52-4165-a83b-436d01b924b6)

**Comment by h9jiang** [comment]
In the next change, we will add a step to tag the pre-release for a given release. 

In my local run, I tried to release a new gopls version 0.16.2. (This is the next patch version so there is no tag related to this release). The flow realize he should tag this release branch's (`gopls-release-branch.0.16`) head commit `gopls/v0.16.2-pre.1` (the pre-release version start from 1). 

My personal credential do not have permission to tag this repo. So I get 403 permission denied error. 

![image](https://github.com/user-attachments/assets/2284d4de-5a00-4abb-8a61-5fe09ccb35c9)

**Comment by h9jiang** [comment]
Next change, the release automation will find the release milestone and create a release tracking issue. All the CL created for this release should reference this issue (part of the commit message).

Happy case 1: release milestone exist, release issue does not exist. A new issue should be created, CL should have this issue number in commit message. 

![image](https://github.com/user-attachments/assets/e5005339-9908-4a15-8bc2-d3484d1b7161)
![image](https://github.com/user-attachments/assets/7f6eea94-1bba-4217-9765-cb49a548de17)

Happy case 2: release milestone exist, release issue exist. All the CL should refer to this pre-existing issue. 

![image](https://github.com/user-attachments/assets/7c49c99e-b444-4662-b2ac-bc036599e64b)
![image](https://github.com/user-attachments/assets/a18cbcab-af9d-4a93-b9eb-1b9c0f847cbe)

Failure case: the milestone is closed. The flow should error out.

![image](https://github.com/user-attachments/assets/e0edc423-2032-468f-90ad-831de49b17bc)

@findleyr  @dmitshur

**Comment by h9jiang** [comment]
Next change, the release automation will re-run a verification after tag the pre-release version. So the workflow will verify the gopls installation twice. From the high level, the flow looks like this: 

```
1. Verify installation using go install gopls@COMMIT (commit is head of release branch)
2. Tag the head of the release branch with gopls.vX.Y.Z-pre.V
3. Verify installation using go install gopls@vX.Y.Z-pre.V (this is tag we just created)
```

In my local relui, a lot of steps are being short circuited. 

![image](https://github.com/user-attachments/assets/b4ee0bfd-384b-418e-9775-05875be869d5)

**Comment by h9jiang** [comment]
Next change, we will rename the current flow to pre-release flow. The pre-release flow is responsible for preparing a pre-release candidate but the release flow is responsible for release a prepared pre-release candidate. 

So a new release will be create to handle the actual release job. This flow will accept a full semantic version string as input in form of `vX.Y.Z-pre.V`. This flow will first validate the input version is semantic correct and logically correct. 

- Is the latest pre-release for its Major.Minor.Patch.
- Has a corresponding tag in the x/tools repo.
- The tag points to the head commit of the release branch.
- The corresponding release tag does not exist in this repo.

Error case, I provide `v0.16.0-pre.3` as input parameter. This flow should error out. Because head of the release branch is way ahead of this tag. 

![image](https://github.com/user-attachments/assets/046e894c-7ddf-4a6f-947d-a07a00cf2afc)

Error case, I provide the `v0.14.2-pre.1` as input parameter, this flow should error out because this specific release is already done. There is a release tag exist in this repo. 

![image](https://github.com/user-attachments/assets/810ed2f3-55cb-4bb9-b0ad-2aead1b17cff)

Unfortunately, the current state of x/tools do not have any happy case I can demonstrate. (no ongoing release like 0.16.3-pre.1). All of the release have both the pre-release tag and release tag.

**Comment by h9jiang** [comment]
Next change: as an improvement, we will replace current approach updating the x/tools dependency using `branch` with using `commit`.  

This is to avoid any possibility where the module proxy provide some outdated info about the branch. 

Normally, this won't be a concern, but in the gopls release, we trigger this `go get` command immediately after the previous CL merged. If we use `go get .... $BRANCH`, it have a higher chance to provide outdated info.

My local relui try to execute the command using commit, and create a auto submit CL.
```
go get -u golang.org/x/tools@66fa8e7351eaa1909b6afcaa95e7adb4e08b1714
```

![image](https://github.com/user-attachments/assets/17a544c9-2968-41be-8d65-f9a4bc7411fe)

![image](https://github.com/user-attachments/assets/9403885b-8ada-4d97-afff-5039f75e6bf9)

**Comment by h9jiang** [comment]
Next change: add a gopls release tag to x/tools release branch. 

The coordinator will provide a semantic version as input to `gopls release flow`. This input version should follow the semantic version schema and the patch version should not be empty. 

The flow will try to find the commit of this pre-release version and tag it with the official release tag. 

E.g. input is `v0.16.0-pre.2`, the flow will find the commit where this tag is pointing to `6c3e23fca26b0bf0433f02b12aec71e3e8d2405b` and tag this commit with release tag `v0.16.0`. 

In my local relui below, the task failed because the tag already exist and I do not permission to tag this repo. :D 

![image](https://github.com/user-attachments/assets/75f10386-4f3c-4ad7-8600-eabaebbaf0da)

**Comment by h9jiang** [comment]
Next change: fix a bug where the cloud run is being triggered against the master branch but the script is being triggered against the release branch. 

![image](https://github.com/user-attachments/assets/970bb0e7-d025-4099-a6d3-9e6f2eee2139)

**Comment by h9jiang** [comment]
Next change: for gopls release, after add the release tag to the x/tools, the last step is to update the dependency for master branch. This step only need to be triggered for minor version release. 

This step is suppose upgrade dependency for x/tool module and it's gopls submodule. The dependency upgrade is independent from each other, so two CL will be created in parallel. 

For local relui test, I short circuit the version validation and release tag. (I just want to test if the dependency works or not). 

The flow execution
![image](https://github.com/user-attachments/assets/ef3943f7-4dda-47ec-ba9c-aaecb9aacb98)

The dependency upgrade for root dir was successful: 
![image](https://github.com/user-attachments/assets/fe231ce1-48d9-4ae4-a500-e369000c2d45)

The dependency upgrade for gopls sub was failing:
![image](https://github.com/user-attachments/assets/64e57e95-12c8-4dea-9bf0-0fadb7c6b002)

The failure is reproducible locally. 

```
/codebase/tools/gopls/ [master] go get -u all
go: downloading github.com/jba/templatecheck v0.7.0
go: downloading github.com/google/safehtml v0.1.0
go: downloading honnef.co/go/tools v0.5.1
go: downloading golang.org/x/vuln v1.1.3
go: downloading mvdan.cc/xurls v1.1.0
go: downloading golang.org/x/telemetry v0.0.0-20240815150606-0693e6240b9b
go: downloading golang.org/x/sys v0.24.0
go: downloading golang.org/x/exp v0.0.0-20240808152545-0cdaa3abc0fa
go: downloading github.com/BurntSushi/toml v1.4.0
go: downloading golang.org/x/exp/typeparams v0.0.0-20240808152545-0cdaa3abc0fa
go: golang.org/x/vuln@v1.1.3 (matching all@upgrade) requires golang.org/x/tools@v0.23.0, not golang.org/x/tools@v0.21.1-0.20240508182429-e35e4ccd0d2d (matching all@upgrade)
go: honnef.co/go/tools@v0.5.1 (matching all@upgrade) requires github.com/BurntSushi/toml@v1.4.1-0.20240526193622-a339e1f7089c, not github.com/BurntSushi/toml@v1.4.0 (matching all@upgrade)
/codebase/tools/gopls/ [master] echo $?                         
1
```

But failure is not within the scope of this CL. So it's working as intended. The purpose of this CL is to upgrade dependency.

**Comment by dmitshur** [comment]
> This step is suppose upgrade dependency for x/tool module and it's gopls submodule. The dependency upgrade is independent from each other, so two CL will be created in parallel.

Have you also considered doing both updates in one CL, and does splitting it out into two parallel CLs end up working out better?

**Comment by h9jiang** [comment]
> > This step is suppose upgrade dependency for x/tool module and it's gopls submodule. The dependency upgrade is independent from each other, so two CL will be created in parallel.
> 
> Have you also considered doing both updates in one CL, and does splitting it out into two parallel CLs end up working out better?

Hi Dmitri, we have some more update on the CL comment. It turns out, we only need to update the gopls sub module in the `x/tools` repo. The release process is outdated. The `x/tools` main module no longer need to be updated. 

But you have a good point, if we ever need to perform two operations, why not just do things together.

**Comment by h9jiang** [comment]
As discussed offline within tooling gophers, we think at the end of the flow, we should send an email announcement. The end state will be: 
- The pre-release flow send an email to tooling gophers so the tooling gophers understand the current progress.
- The release flow send an email to both tooling gophers and go-nuts. 

But right now, we want to do some internal testing first. So both release and pre-release flow will send email to tooling gophers only. At some point in the future, we may also send the final release email to a publicly visible distribution.

@dmitshur @findleyr

**Comment by h9jiang** [comment]
In the next CL, we will introduce the wait for approval step in pre-release flow.

In local relui, I tried to run step after figure out the pre-release version. The release coordinator is suppose to take a look at the version printed in the relui and either approve or deny (which is stop button on the top right).

The `gopls/v0.16.2-pre.1` already exist, so the next pre-release is `gopls/v0.16.2-pre.2`. In this case, I should approve it. 

![image](https://github.com/user-attachments/assets/fa2d3c7b-1ddc-45ed-89cd-ad3cc4e10df5)

Once I approve it, the flow will move forward

![image](https://github.com/user-attachments/assets/0ab15d94-6132-4d07-a956-444b2b851fbc)

**Comment by h9jiang** [comment]
The next CL, we will add a selection input parameter `next version (optional)` to gopls release. The parameter have two options, either `next minor` or `next patch`. 

The UI looks like:

![image](https://github.com/user-attachments/assets/e26f19d9-6cf3-4be8-a5fb-9709a94e9664)

Coordinator will use this field to select which kind of version he/she want to release. But the old input parameter `version` will still be there. The old parameter will also be optional. The `version` will be used for extreme cases only. (you want to release a patch release of a old minor version). 

Most of the case should be handled by the selection instead of the manual input. 

Use the `version (optional)` to release a gopls

![image](https://github.com/user-attachments/assets/e51ac1ea-efdc-4e26-accd-9cbdd65133ee)

Use the `next version (optional)` to release a gopls

![image](https://github.com/user-attachments/assets/a29a1d3f-0fb2-4e95-9d26-61a28c0dab8a)
![image](https://github.com/user-attachments/assets/9adb28ee-65ad-43fc-a52e-beca1f1c40eb)

**Comment by h9jiang** [comment]
We have successfully execute the gopls release and prerelease flow for gopls v0.16.2. So all the implementation necessary is done for this effort. 

Hopefully all the future release will be simple and button clicking!
